### PR TITLE
feat #1  bridge fixes and tests implementation

### PR DESCRIPTION
**Pull request by lsimone**
bridge modification in order to be consistent with the specification and tests implementation.
The tests aim to verify the behavior of atlas widget used through the bridge and the data binding in both directions (from atlas to hsp and vice-versa).
Close #1

### CODE REVIEW

File: public_html/test/atlasWidgets/multiautocomplete/BaseMultiAutoCompleteTestCase.js [hunk 1]
```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright 2013 Amadeus s.a.s.
```
Comment (benouat) [inline public_html/test/atlasWidgets/multiautocomplete/BaseMultiAutoCompleteTestCase.js]: Make sure that the License block you use is correct in all files.
Here the year `2013` is wrong, It should be `2014`

File: public_html/bridge/hsp/atlas.hsp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        var out = new aria.templates.MarkupWriter(tplCtxt);
+        // TODO: manage the content in case of containers (like Div widget)
+// if(this.content && false){
+// var widgInst = lib.processWidgetMarkupBegin(this.name, out, this.config);
+//
+// widgInst.writeMarkupEnd(out);
+// }
+// else{
+            lib.processWidgetMarkup(this.name, out, this.config);
+// }
```
Comment (benouat) [inline public_html/bridge/hsp/atlas.hsp]: Please make sure the formatting and indentation are correct, even when concerned lines are commented out !

File: public_html/bridge/hsp/atlas.hsp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    onConfigChange: function() {
+        this.reloadSection();
+    },
+
+    onNameChange: function() {
+        this.reloadSection();
+    },
+
+    onConfigChange: function() {
+        this.reloadSection();
```
Comment (benouat) [inline public_html/bridge/hsp/atlas.hsp]: This method is duplicated. It's already defined on [Line 77](https://github.com/lsimone/atlas-hashspace-bridge/blob/tests-implementation/public_html/bridge/hsp/atlas.hsp#L77)

File: public_html/bridge/hsp/atlas.hsp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+});
+
+exports.bridge = WidgetBridgeCtrl;
+
+# export template atlas using c:WidgetBridgeCtrl
```
Comment (benouat) [inline public_html/bridge/hsp/atlas.hsp]: We now support a new template definition syntax that you should use

``` cs
{export template atlas using c:WidgetBridgeCtrl}
...
{/template}
```

File: public_html/templates/hsp/page.hsp [hunk 5]
```diff
@@ -15,22 +14,30 @@ var PageCtrl = klass({
...
-        <#widgetbridge wlib="aria.widgets.AriaLib" name="DatePicker" config="{getDatePickerConfig(c)}"/>
+        <#atlas wlib="aria.widgets.AriaLib" name="DatePicker" config="{getDatePickerConfig(c)}"/>
         {c.data.dateValue}
         <hr/>
         @touch:DoubleSlider
-        <#widgetbridge wlib="aria.touch.widgets.TouchWidgetLib" name="DoubleSlider" config="{getDoubleSliderConfig(c)}"/>
+        <#atlas wlib="aria.touch.widgets.TouchWidgetLib" name="DoubleSlider" config="{getDoubleSliderConfig(c)}"/>
         {c.data.sliderValue}
     </div>
 # /template
```
Comment (benouat) [inline public_html/templates/hsp/page.hsp:31]: Same as previous comment.
We now support a new template definition syntax that you should use

``` cs
{export template page using c:PageCtrl}
...
{/template}
```

File: public_html/test/atlasWidgets/TemplateTestCaseWrapper.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Test suite grouping all tests
```
Comment (benouat) [inline public_html/test/atlasWidgets/TemplateTestCaseWrapper.js]: It's apparently not a _Testsuite_ but more a _Testcase_

File: public_html/test/atlasWidgets/iconbutton/IconButton.hsp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+var PageCtrl = klass({
+    attributes:{
+        "moduleCtrl":{type:"object", defaultValue:{}, binding:"2-way"},
+        "data":{type:"object", defaultValue:{}, binding:"2-way"},
+        "id":{type:"object", binding:"2-way"}
+    }
+});
+
+# export template page using c:PageCtrl
```
Comment (benouat) [inline public_html/test/atlasWidgets/iconbutton/IconButton.hsp]: Here again you should change the template definition syntax to

``` cs
{export template page using c:PageCtrl}
...
{/template}
```

File: public_html/test/atlasWidgets/iconbutton/IconButton.tpl [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      'hsp' : 'bridge.aria.widget.BridgeWidgetLib'
+    },
+    $hasScript:true
+}}
+
+    {macro main()}
+         <h2>IconButton</h2>
+         <br />
+
+    {@hsp:Template {
```
Comment (benouat) [inline public_html/test/atlasWidgets/iconbutton/IconButton.tpl]: Indentation is not respected here !

File: public_html/test/atlasWidgets/list/List.hsp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    attributes:{
+        "moduleCtrl":{type:"object", defaultValue:{}, binding:"2-way"},
+        "data":{type:"object", defaultValue:{}, binding:"2-way"},
+        "id":{type:"object", binding:"2-way"}
+    }
+});
+
+# export template page using c:PageCtrl
+    <div class="hsptpl">
+    HASHSPACE OUTER<br/>
```
Comment (benouat) [inline public_html/test/atlasWidgets/list/List.hsp]: Please pay attention to the indentation here !

File: public_html/test/atlasWidgets/list/ListTest.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                        this.assertTrue(this.children.length === 3, "The number of list items is not 3");
+
+                        if (aria.core.Browser.isIE && aria.core.Browser.majorVersion < 9) {
+                            aria.core.Timer.addCallback({
+                                fn : this.waitForIE,
+                                scope : this,
+                                delay : 500
+                            });
+                        } else {
+                            this.addItem();
```
Comment (benouat) [inline public_html/test/atlasWidgets/list/ListTest.js]: This `else` here could be refactored to not be duplicated with the `waitForIE` method

File: public_html/test/atlasWidgets/list/ListTestTpl.tpl [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+
+
+
+    {macro main()}
+
+         <h2>List</h2>
+         <br />
+
+    {@hsp:Template {
```
Comment (benouat) [inline public_html/test/atlasWidgets/list/ListTestTpl.tpl:32]: Indentation should be respected here

File: public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    attributes:{
+        "moduleCtrl":{type:"object", defaultValue:{}, binding:"2-way"},
+        "data":{type:"object", defaultValue:{}, binding:"2-way"},
+        "id":{type:"object", binding:"2-way"}
+    }
+});
+
+# export template page using c:PageCtrl
+    <div class="hsptpl">
+    HASHSPACE OUTER<br/>
```
Comment (benouat) [inline public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp]: Indentation again ! :smile:

File: public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+var atlas = require("bridge/hsp/atlas.hsp").atlas;
+var PageCtrl = klass({
+    attributes:{
+        "moduleCtrl":{type:"object", defaultValue:{}, binding:"2-way"},
+        "data":{type:"object", defaultValue:{}, binding:"2-way"},
+        "id":{type:"object", binding:"2-way"}
+    }
+});
+
+# export template page using c:PageCtrl
```
Comment (benouat) [inline public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp]: Update the template declaration syntax

``` cs
{export template page using c:PageCtrl}
...
{/template}
```

File: public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+        //Bridge to AT widgets
+        <#atlas wlib="aria.widgets.AriaLib" name="MultiAutoComplete" config="{getTextFieldConfig(c)}"/>
+
+    </div>
+# /template
+
+var getTextFieldConfig = function(c) {
+    return {
+            label:"Choose your city: ",
```
Comment (benouat) [inline public_html/test/atlasWidgets/multiautocomplete/MultiAuto.hsp]: Try to be consistent in terms of indentation in the same file.
It seems to be `4` spaces. So try to use `4` everywhere

File: public_html/test/atlasWidgets/multiautocomplete/MultiAutoTpl.tpl [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+
+
+
+    {macro main()}
+
+         <h2>MultiAutoComplete</h2>
+         <br />
+
+    {@hsp:Template {
```
Comment (benouat) [inline public_html/test/atlasWidgets/multiautocomplete/MultiAutoTpl.tpl]: Indentation !!

File: public_html/test/binding/BindingCSS.tpl.css [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    /** Font properties **/
+    {var font = {
+        size : "14pt",
+        weight : "bold",
+        family : "helvetica"
+    } /}
+
+
+    {macro main()}
+                .attpl {
```
Comment (benouat) [inline public_html/test/binding/BindingCSS.tpl.css]: Indentation !!

File: public_html/test/binding/BindingInclude.tpl [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+     'html' : 'aria.html.HtmlLibrary',
+     'hsp' : 'bridge.aria.widget.BridgeWidgetLib'
+  },
+  $hasScript : true
+}}
+
+{macro main()}
+  <div class="attpl">
+
+  ATLAS INNER<br/>
```
Comment (benouat) [inline public_html/test/binding/BindingInclude.tpl]: Indentation !!

File: public_html/test/binding/BindingTemplate.tpl [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    $classpath : "test.binding.BindingTemplate",
+    $wlibs : {
+      'html' : 'aria.html.HtmlLibrary',
+      'hsp' : 'bridge.aria.widget.BridgeWidgetLib'
+    },
+    $hasScript : true,
+    $css: ['test.binding.BindingCSS']
+}}
+    {macro main()}
+TEST ARIA
```
Comment (benouat) [inline public_html/test/binding/BindingTemplate.tpl]: Indentation is not respected here !

File: public_html/test/binding/page.hsp [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+var atlas = require("bridge/hsp/atlas.hsp").atlas;
+var PageCtrl = klass({
+    attributes:{
+        "moduleCtrl":{type:"object", defaultValue:{}, binding:"2-way"},
+        "data":{type:"object", defaultValue:{}, binding:"2-way"},
+        "id":{type:"object", binding:"2-way"}
+    }
+});
+
+# export template page using c:PageCtrl
```
Comment (benouat) [inline public_html/test/binding/page.hsp]: Please update the template declaration syntax to

``` cs
{export template page using c:PageCtrl}
...
{/template}
```

File: public_html/test/binding/page.hsp [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    <div class="hsptpl">
+    HASHSPACE OUTER<br/>
```
Comment (benouat) [inline public_html/test/binding/page.hsp]: Indentation here !!

File: public_html/test/hsbridge/MainTestSuite.js [hunk 21]
```diff
@@ -23,5 +23,10 @@ Aria.classDefinition({
         this.$TestSuite.constructor.call(this);
 
         this.addTests("test.hsbridge.FirstTest");
+        this.addTests("test.binding.BindingTest");
+        this.addTests("test.atlasWidgets.iconbutton.IconButtonTest");
+        this.addTests("test.atlasWidgets.list.ListTest");
+        this.addTests("test.atlasWidgets.multiautocomplete.MultiAutoTest");
+//        this.addTests("test.atlasWidgets.tooltip.TooltipTest");
```
Comment (benouat) [inline public_html/test/hsbridge/MainTestSuite.js]: Keep indentation in-lined, even with commented line.
Make it consistent

File: attester.yml [hunk 22]
```diff
@@ -1,20 +1,20 @@
...
-  - 'node_modules/noder-js/dist/browser'
+  - 'node_modules/hashspace/dist'
  '/test':
   - 'test'
  '/':
   - 'public_html'
 tests:
  aria-templates:
-  bootstrap: '/aria/bootstrap.js'
+  bootstrap: '/aria/ariatemplates-1.6.0-beta.3.js'
```
Comment (benouat) [inline attester.yml:12]: Sounds weird to hardcode AriaTemplates version like that.....
Is there any possibility to **not** do this ?

**Comment by benouat** [comment]
Review is done for e053d3f

Spotted issues are:
- ariatemplates version hardcoded in `attester` configuration file.
- formatting and indentation lacks of consistency.
- hashspace template declaration to be changed from `# template` to `{template}`.
- couple of small refactor.

---

Globally speaking (that means not at all related to this PR code review)
I clearly doubt about the pertinence of testing the `MultiAutoComplete` in that context, knowing that it would be definitely a better solution to rewrite an hashspace version of that widget...

**Comment by lsimone** [comment]
Thank you @benouat 
All the changes have been done, except for ariatemplates version hardcoded in attester configuration file: we should discuss on how to modify it.

**Pull request closed without merging**